src/main/scala/analysis/VSA.scala

```diff
@@ -105,239 +68,97 @@ trait MemoryRegionValueSetAnalysis:
+          mmm.findStackObject(rhs.asInstanceOf[BitVecLiteral].value) match
+            case Some(obj: MemoryRegion) => Some(obj)
+            case _                       => None
         } else {
-            mmm.findObject(lhs.asInstanceOf[BitVecLiteral].value, "data") match
-                case Some(obj: MemoryRegion) =>
-                  return Some(obj)
-                case _ =>
-                return None
-            }
-        None
+          val evaluation: Expr = evaluateExpression(binOp, n, constantProp)
+          if (!evaluation.isInstanceOf[BitVecLiteral]) {
+            return None
+          }
+          mmm.findDataObject(evaluation.asInstanceOf[BitVecLiteral].value) match
+            case Some(obj: MemoryRegion) => Some(obj)
+            case _                       => None
+        }
       case _ =>
         None
   }
 
+  def getValueType(bitVecLiteral: BitVecLiteral): Value = {
+    if (externalFunctions.contains(bitVecLiteral.value)) {
+      LocalAddress(bitVecLiteral, externalFunctions(bitVecLiteral.value))
+    } else if (globals.contains(bitVecLiteral.value)) {
+      GlobalAddress(bitVecLiteral, globals(bitVecLiteral.value))
+    } else if (globalOffsets.contains(bitVecLiteral.value)) {
+      GlobalAddress(bitVecLiteral, resolveGlobalOffset(bitVecLiteral.value))
+    } else if (subroutines.contains(bitVecLiteral.value)) {
+      GlobalAddress(bitVecLiteral, subroutines(bitVecLiteral.value))
+    } else {
+      LiteralValue(bitVecLiteral)
+    }
+  }
 
   /** Default implementation of eval.
-   */
-  def eval(stmt: Statement, s: lattice.sublattice.Element, n: CfgNode): lattice.sublattice.Element = {
-    stmt match {
+    */
+  def eval(cmd: Command, s: lattice.sublattice.Element, n: CfgNode): lattice.sublattice.Element = {
+    Logger.debug(s"eval: $cmd")
+    Logger.debug(s"state: $s")
+    Logger.debug(s"node: $n")
+    cmd match
+      case localAssign: LocalAssign =>
+        localAssign.rhs match
+          case memoryLoad: MemoryLoad =>
+            val region: Option[MemoryRegion] = exprToRegion(memoryLoad.index, n)
+            region match
+              case Some(r: MemoryRegion) =>
+                // this is an exception to the rule and only applies to data regions
+                evaluateExpression(memoryLoad.index, n, constantProp) match
+                  case bitVecLiteral: BitVecLiteral =>
+                    val m = s + (r -> Set(getValueType(bitVecLiteral)))
+                    m + (localAssign.lhs -> m(r))
+                  case _ =>
+                    s + (localAssign.lhs -> s(r))
+              case None =>
+                Logger.warn("could not find region for " + localAssign)
+                s
+          case e: Expr => {
+            val evaled = evaluateExpression(e, n, constantProp)
+            evaled match
+              case bv: BitVecLiteral => s + (localAssign.lhs -> Set(getValueType(evaled.asInstanceOf[BitVecLiteral])))
```

Comment:
Should just use `bv` instead of `evaled.asInstanceOf[BitVecLiteral]`

Reply:
Fixed [Upcoming merge from yousif-memory-region-analysis]